### DISCUSSION TITLE
XPath2Evaluate not working

### DISCUSSION THREAD

**Discussion by A714m35** [StefH/XPath2.Net#68] [General]
Hi,

I have a path expression such as /path1/path2/path3[path4 and path5/path6=path2/path3] . I cannot have a result from this kind of expression, could you kindly help me understand how to make it work or make a fix for this, please? 

Very Grateful @StefH @kashifsoofi @martin-honnen

**Comment by martin-honnen** [comment]
Consider to post a minimal but complete sample allowing us to understand and reproduce the problem. Do you get an error? If so, which one? Or are you simply not selecting the nodes you are looking for? Have you tried another engine to check whether the problem is with your XPath expression or with XPath2? Anyway, the XPath sample alone doesn't help, show us the XML and the code to reproduce the problem.

**Reply by A714m35** [reply]
Thank you so much for answering me (: @martin-honnen 

I will share an xml and the path that is failing.


Very grateful,
Joana

**Comment by martin-honnen** [comment]
I don't see which sense `sum(/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'O' and ram:RateApplicablePercent=ram:ApplicableTradeTax/ram:RateApplicablePercent]/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount)1010` with the 1010 at the end is supposed to make.
Any, post a sample that allows us to reproduce the error you get and tell us exactly which error you get.

**Reply by A714m35** [reply]
@martin-honnen  my bad ,  my apologies

thanks for the quick answer, you are very kind.


Very grateful,
Joana :)

**Reply by martin-honnen** [reply]
The XPath seems XPath 1.0, can't you just use the built-in .NET XPath 1.0 support? Doesn't that also give -5?

**Reply by A714m35** [reply]
yes

**Reply by martin-honnen** [reply]
Help you achieve what? Do you think there is a bug in the XPath2 library? Then provide evidence of that with a minimal but complete test case.
If the problem is just that your XPath expression doesn't select or compute what you expect with any XPath library/implementation then I don't see this Github discussion issue as the right place to get you help with finding the right XPath.

**Reply by A714m35** [reply]
You are right.
Thank you very much for your time, kindness and patience.

Best regards

**Comment by nverwer** [comment]
This is **not** a problem with XPath2.net. I tried it using Saxon, and that also gives -5 as the result.

I will give you a hint anyway:

You say that you are executing this rule in a context, but when you start an expression with `/rsm:CrossIndustryInvoice`, the context is not used because your XPath starts at the document root.

I would suggest splitting up the expression, which I did using XQuery, like

```
declare namespace qdt="urn:un:unece:uncefact:data:standard:QualifiedDataType:100";
declare namespace ram="urn:un:unece:uncefact:data:standard:ReusableAggregateBusinessInformationEntity:100";
declare namespace rsm="urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100";
declare namespace udt="urn:un:unece:uncefact:data:standard:UnqualifiedDataType:100";

let $amounts1 := /rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'O' and ram:RateApplicablePercent=ram:ApplicableTradeTax/ram:RateApplicablePercent]/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount
let $part1 := round(sum($amounts1) * 10 * 10) div 100
let $amounts2 := /rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='true' and ram:CategoryTradeTax/ram:CategoryCode='O' and ram:RateApplicablePercent=ram:CategoryTradeTax/ram:RateApplicablePercent]/ram:ActualAmount
let $part2 := (round(sum($amounts2) * 10 * 10) div 100)
let $amounts3 := /rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:ApplicableHeaderTradeSettlement/ram:SpecifiedTradeAllowanceCharge[ram:ChargeIndicator/udt:Indicator='false' and ram:CategoryTradeTax/ram:CategoryCode='O']/ram:ActualAmount
let $part3 := (round(sum($amounts3) * 10 *10) div 100)
let $total := $part1 + $part2 - $part3
return (count($amounts1), $part1, count($amounts2), $part2, $part3, $total)
```

The result is `0 0 0 0 5 -5`, so the first two parts have empty sums which add up to zero.

In the first part, the `ram:RateApplicablePercent` you are testing does not exist, so you are testing `() = ()`, which is false. This causes zero terms to be summed.
Maybe you meant

```
/rsm:CrossIndustryInvoice/rsm:SupplyChainTradeTransaction/ram:IncludedSupplyChainTradeLineItem/ram:SpecifiedLineTradeSettlement[ram:ApplicableTradeTax/ram:CategoryCode = 'O' and ( empty(ram:RateApplicablePercent) or  ram:RateApplicablePercent=ram:ApplicableTradeTax/ram:RateApplicablePercent )]/ram:SpecifiedTradeSettlementLineMonetarySummation/ram:LineTotalAmount
```

In that case, it would still be better to split this expression in meaningful parts. I have not worked with XPath2.net for some time, so I don't know how to do that in a neat way. It depends on the context of your program, the programming language you use, etcetera. Maybe you could use [LINQ](https://learn.microsoft.com/en-us/dotnet/standard/linq/linq-xml-overview) instead.

**Reply by A714m35** [reply]
@nverwer thank you so much. It was my bad, I then realized it. My apologies.

You were very kind and helpful, I'm very  grateful.

Best regards,
Joana (:
